1. Create the Prediciton Probability "Error" results for the `xgboost` and `ffnn` models analagously to the `transformer` model produced above.

In [19]:
import numpy as np
import pandas as pd
import plotly.express as px
import scipy.stats as stats
df_indicators = pd.read_csv('country_indicators.csv')
df_preds = pd.read_csv('test_predictions.csv')



In [20]:
#Prediciton Probability "Error" results for the `xgboost` model
df_preds['xgboost_probability_prediction_error'] = np.abs(df_preds['y_true_xgboost'].astype(float) - df_preds['y_pred_proba_xgboost'])
df_preds[['y_true_xgboost','y_pred_proba_xgboost','xgboost_probability_prediction_error']]

y_true_xgboost  y_pred_proba_xgboost  \
0             False              0.066500   
1             False              0.099643   
2              True              0.704086   
3              True              0.638444   
4             False              0.608380   
..              ...                   ...   
359           False              0.079453   
360           False              0.060189   
361            True              0.697625   
362           False              0.729246   
363           False              0.591722   

     xgboost_probability_prediction_error  
0                                0.066500  
1                                0.099643  
2                                0.295914  
3                                0.361556  
4                                0.608380  
..                                    ...  
359                              0.079453  
360                              0.060189  
361                              0.302375  
362                              0.729246  
363                              0.591722  

[364 rows x 3 columns]

In [21]:
#Prediciton Probability "Error" results for the `ffnn` model
df_preds['ffnn_probability_prediction_error'] = np.abs(df_preds['y_true_ffnn'].astype(float) - df_preds['y_pred_proba_ffnn'])
df_preds[['y_true_ffnn','y_pred_proba_ffnn','ffnn_probability_prediction_error']]

y_true_ffnn  y_pred_proba_ffnn  ffnn_probability_prediction_error
0          False           0.409958                           0.409958
1          False           0.406696                           0.406696
2          False           0.545236                           0.545236
3          False           0.534560                           0.534560
4           True           0.538583                           0.461417
..           ...                ...                                ...
359        False           0.291874                           0.291874
360        False           0.300321                           0.300321
361        False           0.335496                           0.335496
362        False           0.324000                           0.324000
363         True           0.332455                           0.667545

[364 rows x 3 columns]

2. Create a bootsrap confidence interval for the average Prediction Probability "Error" for one of these models using all the data.

In [22]:
n = 5
repititions = 5000
transformer_stat = np.zeros(repititions)
sample_1 = np.arange(n)

# bootstrapping a single sample
for i in range(repititions):
    bootstrap_sample_transformer = np.random.choice(sample_1, size=n, replace = True)
    transformer_stat[i] = bootstrap_sample_transformer.mean()

np.quantile(transformer_stat,[0.025, 0.975])

array([0.8, 3.2])

3. Create the Prediction Classification "Correctness" results of "correct" and "incorrect" predictions for the `transformer`, `xgboost` and `ffnn` models; or, an alternative "either/or" breakdown of interest (such as "wrongly predicted no escalation" versus all the other categories combined).

In [23]:
#Prediction Classification "Correctness" results of "correct" and "incorrect" predictions for the `transformer` model
threshold = 0.5 

df_preds['transformer_classifcation_performance_outcome'] = None

tmp = df_preds['transformer_classifcation_performance_outcome'].copy()
TP_pos_pred_correct = df_preds.y_true_transformer & (df_preds.y_pred_proba_transformer>threshold)
tmp[TP_pos_pred_correct] = "correctly predicted escalation"
TN_neg_pred_correct = (~df_preds.y_true_transformer) & (df_preds.y_pred_proba_transformer<=threshold)
tmp[TN_neg_pred_correct] = "correctly predicted no escalation"
FP_pos_pred_wrong = (~df_preds.y_true_transformer) & (df_preds.y_pred_proba_transformer>threshold)
tmp[FP_pos_pred_wrong] = "wrongly predicted escalation"
FN_neg_pred_wrong = df_preds.y_true_transformer & (df_preds.y_pred_proba_transformer<=threshold)
tmp[FN_neg_pred_wrong] = "wrongly predicted no escalation"

df_preds['transformer_classifcation_performance_outcome'] = tmp
df_preds[['y_true_transformer','y_pred_transformer','transformer_classifcation_performance_outcome']][30:50]

y_true_transformer  y_pred_transformer  \
30                True                True   
31               False                True   
32               False                True   
33               False                True   
34               False               False   
35               False                True   
36               False               False   
37               False               False   
38               False                True   
39               False                True   
40               False                True   
41               False                True   
42               False                True   
43               False               False   
44               False               False   
45               False               False   
46               False               False   
47                True               False   
48               False               False   
49               False               False   

   transformer_classifcation_performance_outcome  
30                correctly predicted escalation  
31                  wrongly predicted escalation  
32                  wrongly predicted escalation  
33                  wrongly predicted escalation  
34             correctly predicted no escalation  
35                  wrongly predicted escalation  
36             correctly predicted no escalation  
37             correctly predicted no escalation  
38                  wrongly predicted escalation  
39                  wrongly predicted escalation  
40                  wrongly predicted escalation  
41                  wrongly predicted escalation  
42                  wrongly predicted escalation  
43             correctly predicted no escalation  
44             correctly predicted no escalation  
45             correctly predicted no escalation  
46             correctly predicted no escalation  
47               wrongly predicted no escalation  
48             correctly predicted no escalation  
49             correctly predicted no escalation

In [24]:
#Prediction Classification "Correctness" results of "correct" and "incorrect" predictions for the `xgboost` model
threshold = 0.5 

df_preds['xgboost_classifcation_performance_outcome'] = None

xmp = df_preds['xgboost_classifcation_performance_outcome'].copy()
xP_pos_pred_correct = df_preds.y_true_xgboost & (df_preds.y_pred_proba_xgboost>threshold)
xmp[xP_pos_pred_correct] = "correctly predicted escalation"
xN_neg_pred_correct = (~df_preds.y_true_xgboost) & (df_preds.y_pred_proba_xgboost<=threshold)
xmp[xN_neg_pred_correct] = "correctly predicted no escalation"
FP_pos_pred_wrong = (~df_preds.y_true_xgboost) & (df_preds.y_pred_proba_xgboost>threshold)
xmp[FP_pos_pred_wrong] = "wrongly predicted escalation"
FN_neg_pred_wrong = df_preds.y_true_xgboost & (df_preds.y_pred_proba_xgboost<=threshold)
xmp[FN_neg_pred_wrong] = "wrongly predicted no escalation"

df_preds['xgboost_classifcation_performance_outcome'] = xmp
df_preds[['y_true_xgboost','y_pred_xgboost','xgboost_classifcation_performance_outcome']][30:50]

y_true_xgboost  y_pred_xgboost xgboost_classifcation_performance_outcome
30           False            True              wrongly predicted escalation
31           False            True              wrongly predicted escalation
32           False            True              wrongly predicted escalation
33           False            True              wrongly predicted escalation
34           False           False         correctly predicted no escalation
35           False            True              wrongly predicted escalation
36           False            True              wrongly predicted escalation
37           False            True              wrongly predicted escalation
38            True            True            correctly predicted escalation
39           False            True              wrongly predicted escalation
40           False            True              wrongly predicted escalation
41            True            True            correctly predicted escalation
42           False            True              wrongly predicted escalation
43           False            True              wrongly predicted escalation
44           False            True              wrongly predicted escalation
45            True            True            correctly predicted escalation
46            True           False           wrongly predicted no escalation
47           False            True              wrongly predicted escalation
48            True            True            correctly predicted escalation
49           False            True              wrongly predicted escalation

In [25]:
#Prediction Classification "Correctness" results of "correct" and "incorrect" predictions for the `ffnn` model
threshold = 0.5 

df_preds['ffnn_classifcation_performance_outcome'] = None

fmp = df_preds['ffnn_classifcation_performance_outcome'].copy()
fP_pos_pred_correct = df_preds.y_true_ffnn & (df_preds.y_pred_proba_ffnn>threshold)
fmp[fP_pos_pred_correct] = "correctly predicted escalation"
fN_neg_pred_correct = (~df_preds.y_true_ffnn) & (df_preds.y_pred_proba_ffnn<=threshold)
fmp[fN_neg_pred_correct] = "correctly predicted no escalation"
FP_pos_pred_wrong = (~df_preds.y_true_ffnn) & (df_preds.y_pred_proba_ffnn>threshold)
fmp[FP_pos_pred_wrong] = "wrongly predicted escalation"
FN_neg_pred_wrong = df_preds.y_true_ffnn & (df_preds.y_pred_proba_ffnn<=threshold)
fmp[FN_neg_pred_wrong] = "wrongly predicted no escalation"

df_preds['ffnn_classifcation_performance_outcome'] = fmp
df_preds[['y_true_ffnn','y_pred_ffnn','ffnn_classifcation_performance_outcome']][30:50]

y_true_ffnn  y_pred_ffnn ffnn_classifcation_performance_outcome
30         True         True         correctly predicted escalation
31        False        False      correctly predicted no escalation
32        False         True           wrongly predicted escalation
33        False         True           wrongly predicted escalation
34        False         True           wrongly predicted escalation
35        False         True           wrongly predicted escalation
36        False         True           wrongly predicted escalation
37        False        False      correctly predicted no escalation
38        False        False      correctly predicted no escalation
39        False        False      correctly predicted no escalation
40        False        False      correctly predicted no escalation
41        False        False      correctly predicted no escalation
42        False        False      correctly predicted no escalation
43        False        False      correctly predicted no escalation
44        False        False      correctly predicted no escalation
45        False        False      correctly predicted no escalation
46        False        False      correctly predicted no escalation
47         True        False        wrongly predicted no escalation
48        False         True           wrongly predicted escalation
49        False         True           wrongly predicted escalation

4. Perform a one sample hypothesis test of the proportion of a specific Prediction Classification "Correctness" category for another of these models using all the data.


In [26]:
#Analysis of Prediction Error in xgboost Model
df_preds['xgboost_probability_prediction_error'] = np.abs(df_preds['y_true_xgboost'].astype(float) - df_preds['y_pred_proba_xgboost'])
df_preds[['y_true_xgboost','y_pred_proba_xgboost','xgboost_probability_prediction_error']]

y_true_xgboost  y_pred_proba_xgboost  \
0             False              0.066500   
1             False              0.099643   
2              True              0.704086   
3              True              0.638444   
4             False              0.608380   
..              ...                   ...   
359           False              0.079453   
360           False              0.060189   
361            True              0.697625   
362           False              0.729246   
363           False              0.591722   

     xgboost_probability_prediction_error  
0                                0.066500  
1                                0.099643  
2                                0.295914  
3                                0.361556  
4                                0.608380  
..                                    ...  
359                              0.079453  
360                              0.060189  
361                              0.302375  
362                              0.729246  
363                              0.591722  

[364 rows x 3 columns]

In [27]:
#Analysis of Prediction Error in ffnn Model
df_preds['ffnn_probability_prediction_error'] = np.abs(df_preds['y_true_ffnn'].astype(float) - df_preds['y_pred_proba_ffnn'])
df_preds[['y_true_ffnn','y_pred_proba_ffnn','ffnn_probability_prediction_error']]

y_true_ffnn  y_pred_proba_ffnn  ffnn_probability_prediction_error
0          False           0.409958                           0.409958
1          False           0.406696                           0.406696
2          False           0.545236                           0.545236
3          False           0.534560                           0.534560
4           True           0.538583                           0.461417
..           ...                ...                                ...
359        False           0.291874                           0.291874
360        False           0.300321                           0.300321
361        False           0.335496                           0.335496
362        False           0.324000                           0.324000
363         True           0.332455                           0.667545

[364 rows x 3 columns]

## ------------------------ The correlation between xgboost & ffnn---------------------------
> $ H_{0} $: The proportion of correct predictions for the xgboost model is equal to the proportion of correct predictions for the ffnn model.

> $ H_{1} $: The proportion of correct predictions for the xgboost model is not equal to the proportion of correct predictions for the ffnn model.

In [28]:
# Set the random seed for reproducibility.
np.random.seed(130)

# Define the number of permutations (bootstrap samples) to generate.
num_permutations = 10000

# Create an array to store simulated differences between models' performance.
simulated_differences = np.array([0.0] * num_permutations, dtype=float)

# Extract the data for XGBoost: y_true_xgboost represents the "Correctness" category.
# Extract the data for FFNN: y_true_ffnn represents the "Correctness" category.
for i in range(num_permutations):
    # Generate bootstrapped samples for XGBoost and FFNN separately.
    # Randomly select indices with replacement to create bootstrapped samples for XGBoost and FFNN.
    xgboost_sampled_idx = df_preds['y_true_xgboost'].sample(frac=1, replace=True).index
    ffnn_sampled_idx = df_preds['y_true_ffnn'].sample(frac=1, replace=True).index

    # Create bootstrapped datasets for XGBoost and FFNN based on the sampled indices.
    xgboost_bootstrapped = df_preds.loc[xgboost_sampled_idx].copy()
    ffnn_bootstrapped = df_preds.loc[ffnn_sampled_idx].copy()

    # Calculate the mean of the "Correctness" category for XGBoost and FFNN bootstrapped samples.
    xgboost_mean = xgboost_bootstrapped['y_true_xgboost'].mean()
    ffnn_mean = ffnn_bootstrapped['y_true_ffnn'].mean()

    # Calculate the difference in means for the bootstrapped samples.
    simulated_differences[i] = xgboost_mean - ffnn_mean

# Calculate the observed test statistic, representing the actual difference in means between XGBoost and FFNN.
observed_test_statistic = df_preds['y_true_xgboost'].mean() - df_preds['y_true_ffnn'].mean()

# Define the null hypothesis (H0): Assuming no difference between the models, the difference in means should be 0.
null_hypothesis = 0

# Calculate the number of simulated differences as extreme or more extreme than the observed test statistic.
num_as_or_more_extreme = (np.abs(np.array(simulated_differences) - null_hypothesis) >= np.abs(observed_test_statistic - null_hypothesis)).sum()

# Calculate the p-value by determining the proportion of extreme simulated differences.
p_value = num_as_or_more_extreme / num_permutations

# Output the calculated p-value.
p_value

0.9617

#### Summary:
> Comparison of Prediction Accuracy Between xgboost and ffnn Models: Lack of significant Differences: With a calculated p-value of 0.9617, the analysis indicates that there is insufficient evidence to reject the null hypothesis. This suggests that there is no significant difference in the proportions of correct predictions between the XGBoost and FFNN models. The high p-value, which exceeds the commonly chosen significance level (alpha) of 0.05, underscores the absence of statistical evidence to support the notion that one model outperforms the other in terms of prediction accuracy.

5. Consider the "Progress Indicators" data and use "boolean selection" with one (or more) of the columns to restrict the data to a subset (of rows) of data and repeat either of the (confidence interval and hypothesis testing) analyses above but this time instead only using this specified subset of countries.

In [29]:
# Filter the dataset based on specific criteria for "fsi_e3:_human_flight_and_brain_drain."
# Select countries with values greater than or equal to 5.0 in this column.
subset_data = df_indicators[df_indicators['fsi_e3:_human_flight_and_brain_drain'] >= 5.0]

# Calculate the mean and standard deviation for the selected subset of data.
mean = subset_data['fsi_e3:_human_flight_and_brain_drain'].mean()
std_dev = subset_data['fsi_e3:_human_flight_and_brain_drain'].std()
sample_size = len(subset_data)

# Calculate the confidence interval at a 95% confidence level.
confidence_level = 0.95
z_score = stats.norm.ppf((1 + confidence_level) / 2)
margin_of_error = z_score * (std_dev / (sample_size ** 0.5))

# Define the confidence interval using the calculated mean and margin of error.
confidence_interval = (mean - margin_of_error, mean + margin_of_error)

# Print the calculated confidence interval.
print("Confidence Interval:", confidence_interval)

# Create a bar chart to visualize the confidence interval for "fsi_e3:_human_flight_and_brain_drain."
fig = px.bar(x=['Mean'], y=[mean], error_y=[margin_of_error], title="Confidence Interval for fsi_e3:_human_flight_and_brain_drain")
fig.update_traces(marker_color='blue', marker_line_color='black', marker_line_width=1, error_x_thickness=1, error_x_color='black')
fig.show()


Confidence Interval: (6.414838394483165, 6.83268635799208)


#### Summary:
> We filtered the data to focus on countries with "fsi_e3:_human_flight_and_brain_drain" values greater than or equal to 5.0. Within this subset, we calculated the mean value, which came out to be approximately 6.6237. Additionally, we established a 95% confidence interval for this subset, suggesting that we can be reasonably confident that the true population mean falls within the range of 6.4148 to 6.8327. 

In [30]:
# Filter the dataset based on specific criteria for "fsi_e2:_economic_inequality."
# Select countries with values greater than or equal to 5.0 in this column.
subset_data = df_indicators[df_indicators['fsi_e2:_economic_inequality'] >= 5.0]

# Calculate the mean and standard deviation for the selected subset of data.
mean = subset_data['fsi_e2:_economic_inequality'].mean()
std_dev = subset_data['fsi_e2:_economic_inequality'].std()
sample_size = len(subset_data)

# Calculate the confidence interval at a 95% confidence level.
confidence_level = 0.95
z_score = stats.norm.ppf((1 + confidence_level) / 2)
margin_of_error = z_score * (std_dev / (sample_size ** 0.5))

# Define the confidence interval using the calculated mean and margin of error.
confidence_interval = (mean - margin_of_error, mean + margin_of_error)

# Print the calculated confidence interval.
print("Confidence Interval:", confidence_interval)

# Create a bar chart to visualize the confidence interval for "fsi_e2:_economic_inequality."
fig = px.bar(x=['Mean'], y=[mean], error_y=[margin_of_error], title="Confidence Interval for fsi_e2:_economic_inequality")
fig.update_traces(marker_color='blue', marker_line_color='black', marker_line_width=1, error_x_thickness=1, error_x_color='black')
fig.show()


Confidence Interval: (6.672001828028841, 7.1540851284928975)


#### Summary:
> We filtered the data to focus on countries with "fsi_e2:_economic_inequality" values greater than or equal to 5.0. Within this subset, we calculated the mean value, which was approximately 6.9131. Additionally, we established a 95% confidence interval for this subset, suggesting that we can be reasonably confident that the true population mean falls within the range of 6.6720 to 7.1541. This interval provides us with a level of confidence about the economic inequality within the selected countries and helps to quantify the uncertainty associated with our estimate.

6. (and 7.) Create a two-sample bootstrap confidence interval and perform a hypothesis test comparing the performance of a single model for the data subset created above versus the remaining data not included in that data subset.


## Evaluating Economic Inequality Differences between Countries with Varied Human Flight and Brain Drain Scores
> $ H_{0}$:
The mean economic inequality (measured by "fsi_e2:_economic_inequality") is the same for countries with a "fsi_e3:_human_flight_and_brain_drain" score greater than or equal to 8.0 and countries with a score less than 8.0. In other words, there is no difference in economic inequality between these two groups of countries.

> $ H_{1}$:
The mean economic inequality is different for countries with a "fsi_e3:_human_flight_and_brain_drain" score greater than or equal to 8.0 and countries with a score less than 8.0. There is a significant difference in economic inequality between these two groups of countries.

In [31]:
# Filter the dataset to create a subset based on "fsi_e3:_human_flight_and_brain_drain" criteria.
subset_data = df_indicators[df_indicators['fsi_e3:_human_flight_and_brain_drain'] >= 8.0]

# Create the remaining dataset not included in the subset.
remaining_data = df_indicators[df_indicators['fsi_e3:_human_flight_and_brain_drain'] < 8.0]

# Define the number of bootstrap samples to generate.
n_samples = 10000

# Initialize an array to store the mean differences.
mean_differences = np.empty(n_samples)

# Perform bootstrapping to analyze differences in "fsi_e2:_economic_inequality."
for i in range(n_samples):
    # Create bootstrap samples for the subset and remaining data separately.
    subset_sample = np.random.choice(subset_data['fsi_e2:_economic_inequality'], len(subset_data), replace=True)
    remaining_sample = np.random.choice(remaining_data['fsi_e2:_economic_inequality'], len(remaining_data), replace=True)
    mean_differences[i] = np.mean(subset_sample) - np.mean(remaining_sample)

# Calculate the 95% confidence interval for the differences in means.
confidence_interval = np.percentile(mean_differences, [2.5, 97.5])

# Calculate the observed mean difference between the subset and remaining data.
observed_mean_difference = np.mean(subset_data['fsi_e2:_economic_inequality']) - np.mean(remaining_data['fsi_e2:_economic_inequality'])

# Calculate the p-value for the observed mean difference.
p_value = np.mean(mean_differences >= observed_mean_difference)

# Print the results and create a histogram plot with a vertical line for the observed mean difference.
print("95% Confidence Interval:", confidence_interval)
print("Observed Mean Difference:", observed_mean_difference)
print("p-value:", p_value)

# Create a figure for a histogram plot.
fig = px.histogram(mean_differences, nbins=30, labels={'value': 'Mean Differences'}, title='Bootstrap Analysis of "fsi_e2:_economic_inequality"')

# Add a vertical line for the observed mean difference.
fig.add_shape(
    type='line',
    x0=observed_mean_difference,
    x1=observed_mean_difference,
    y0=0,
    y1=0.2,  # Adjust the value for the y-range
    line=dict(color='red', width=2, dash='dash')
)

# Add shaded regions for the confidence interval.
fig.add_shape(
    type='rect',
    x0=confidence_interval[0],
    x1=confidence_interval[1],
    y0=0,
    y1=0.2,  # Adjust the value for the y-range
    line=dict(color='gray'),
    fillcolor='gray',
    opacity=0.3
)

# Show the plot.
fig.show()


95% Confidence Interval: [-0.44569099  1.77704193]
Observed Mean Difference: 0.6347826086956543
p-value: 0.489


#### Summary:
> In this analysis, a 95% confidence interval was computed to assess the difference in economic inequality between countries with high and low scores on the "Human Flight and Brain Drain" indicator. The confidence interval ranged from -0.462 to 1.768. The observed mean difference in economic inequality between these two groups of countries was 0.635. 
The p-value obtained was 0.495, suggesting that there is no statistically significant difference in economic inequality between countries with high and low scores on the "Human Flight and Brain Drain" indicator. This means that the observed mean difference is not significantly different from what we would expect to see by random chance alone.
In summary, the data does not provide strong evidence to conclude that there is a significant difference in economic inequality between these two groups of countries based on the "Human Flight and Brain Drain" indicator.

In [32]:
# Filter the dataset based on "fsi_e1:_economy" criteria.
subset_data = df_indicators[df_indicators['fsi_e1:_economy'] >= 8.0]

# Create a dataset with countries that do not meet the specified criteria.
remaining_data = df_indicators[df_indicators['fsi_e1:_economy'] < 8.0]

# Define the number of bootstrap samples to generate.
n_samples = 10000

# Initialize an array to store the differences in means.
mean_differences = np.empty(n_samples)

# Perform bootstrapping to analyze differences in "fsi_p2:_public_services."
for i in range(n_samples):
    # Generate bootstrap samples for both the subset and remaining data separately.
    subset_sample = np.random.choice(subset_data['fsi_p2:_public_services'], len(subset_data), replace=True)
    remaining_sample = np.random.choice(remaining_data['fsi_p2:_public_services'], len(remaining_data), replace=True)
    mean_differences[i] = np.mean(subset_sample) - np.mean(remaining_sample)

# Calculate the 95% confidence interval for the differences in means.
confidence_interval = np.percentile(mean_differences, [2.5, 97.5])

# Calculate the observed mean difference between the subset and remaining data.
observed_mean_difference = np.mean(subset_data['fsi_p2:_public_services']) - np.mean(remaining_data['fsi_p2:_public_services'])

# Calculate the p-value for the observed mean difference.
p_value = np.mean(mean_differences >= observed_mean_difference)

# Print the results.

# Create a histogram plot with annotations.
fig = px.histogram(mean_differences, nbins=30, labels={'value': 'Mean Differences'}, title='Bootstrap Analysis of "fsi_e1:_economy" and "fsi_p2:_public_services"')

# Add a vertical line for the observed mean difference.
fig.add_shape(
    type='line',
    x0=observed_mean_difference,
    x1=observed_mean_difference,
    y0=0,
    y1=0.2,  # Adjust the value for the y-range
    line=dict(color='red', width=2, dash='dash')
)

# Add shaded regions for the confidence interval.
fig.add_shape(
    type='rect',
    x0=confidence_interval[0],
    x1=confidence_interval[1],
    y0=0,
    y1=0.2,  # Adjust the value for the y-range
    line=dict(color='gray'),
    fillcolor='gray',
    opacity=0.3
)

# Show the plot.
fig.show()


#### Summary:
>For the comparison between "fsi_e1:_economy" and "fsi_p2:_public_services," here are the key findings: (1) A 95% confidence interval for the mean difference suggests the true difference lies between approximately 2.83 and 4.06. (2) The observed mean difference is 3.47, indicating higher "fsi_e1:_economy" scores correspond to a 3.47-point higher mean in "fsi_p2:_public_services." (3) However, the p-value of 0.4992 implies no statistically significant difference. This suggests the observed difference could be due to random variation, making it inconclusive. In summary, there's no strong evidence that "fsi_e1:_economy" significantly impacts "fsi_p2:_public_services" in this dataset.

8. (and 9.) Create a bootstrap confidence interval and a hypothesis test comparing the performance of two the models across all the data on the basis of a "paired" sample analysis (by transforming the paired sample into a single $z_i=x_i-y_i$ difference sample).

In [33]:
sample_xgboost_error = df_preds['xgboost_probability_prediction_error']
sample_ffnn_error = df_preds['ffnn_probability_prediction_error']
observed_test_statistics = sample_xgboost_error - sample_ffnn_error
n_bootstrap_samples = 1000

bootstrap_sample_means = np.zeros(n_bootstrap_samples)

for i in range(n_bootstrap_samples):
  
    bootstrap_sample = np.random.choice(observed_test_statistics, size=len(observed_test_statistics), replace=True)
    
    
    bootstrap_sample_means[i] = np.mean(bootstrap_sample)


confidence_interval = np.quantile(bootstrap_sample_means, [.25, .975])
confidence_interval

array([0.02214184, 0.0478871 ])

In [34]:
sample_xgboost_error = df_preds['xgboost_probability_prediction_error']
sample_ffnn_error = df_preds['ffnn_probability_prediction_error']
sample_difference = sample_xgboost_error - sample_ffnn_error
R,n=1000, len(sample_xgboost_error)
mean_diff= observed_test_statistics.mean()
mean_diffs= np.zeros(R)
for rep in range(R):
    order_of_subtraction = np.random.choice([1,-1], p=[0.5,0.5], size=n)
    sample_difference_random_order = sample_difference*order_of_subtraction
    mean_diffs[rep]=sample_difference_random_order.mean()
    
p = np.mean(np.abs(mean_diffs) >= np.abs(mean_diff))


p

0.01

10. Repeat the above analyses for different model pairs on some different subsets of data.

In [35]:
import pandas as pd
import numpy as np
from scipy.stats import ttest_rel

country_indicators = pd.read_csv("country_indicators.csv")
test_predictions = pd.read_csv("test_predictions.csv")

test_predictions['xgboost_probability_prediction_error'] = np.abs(df_preds['y_true_xgboost'].astype(float) - df_preds['y_pred_proba_xgboost'])
test_predictions['ffnn_probability_prediction_error'] = np.abs(df_preds['y_true_ffnn'].astype(float) - df_preds['y_pred_proba_ffnn'])
# country == alert and country == stable
subset_alert = country_indicators[country_indicators['fsi_category'] == 'Alert']
subset_stable = country_indicators[country_indicators['fsi_category'] == 'Stable']

# data of country==alert and data of country== stable
data_subset_alert = test_predictions[test_predictions['iso3'].isin(subset_alert['iso3'])]
data_subset_stable = test_predictions[test_predictions['iso3'].isin(subset_stable['iso3'])]

# ttest
ttest_statistic, alert_ttest_pvalue = ttest_rel(data_subset_alert['xgboost_probability_prediction_error'], data_subset_alert['ffnn_probability_prediction_error'])


ttest_statistic, stable_ttest_pvalue = ttest_rel(data_subset_stable['xgboost_probability_prediction_error'], data_subset_stable['ffnn_probability_prediction_error'])
print(alert_ttest_pvalue, stable_ttest_pvalue)

0.05418064493142108 0.04640297000427797
